relaymgr/relaymgr.go

```diff
@@ -0,0 +1,138 @@
+func New(ctx context.Context, h host.Host, relayMultiaddress string) (*RelayManager, error) {
+	relayAddr, err := multiaddr.NewMultiaddr(relayMultiaddress)
+	if err != nil {
+		return nil, fmt.Errorf("parsing relay multiaddr: %s", err)
+	}
+	addrInfo, err := peer.AddrInfoFromP2pAddr(relayAddr)
+	if err != nil {
+		return nil, fmt.Errorf("get addr-info from relay multiaddr: %s", err)
+	}
+
+	closeCtx, closeSignal := context.WithCancel(context.Background())
+	rm := &RelayManager{
+		host:      h,
+		relayAddr: *addrInfo,
+
+		closeCtx:    closeCtx,
+		closeSignal: closeSignal,
+		closed:      make(chan struct{}),
+	}
+	rm.connNotifee = &connNotifee{rm: rm}
+	h.Network().Notify(rm.connNotifee)
+
+	if err := rm.connect(); err != nil {
+		return nil, fmt.Errorf("connecting to relay: %s", err)
+	}
+
+	go rm.keepHealthy()
+
+	return rm, nil
+}
+
+// Close stops relay manager work to keep a healthy connection with the relay.
+func (rm *RelayManager) Close() error {
+	rm.closeOnce.Do(func() {
+		log.Infof("closing relay manager")
+
+		rm.host.ConnManager().Unprotect(rm.relayAddr.ID, connProtectTag)
+		rm.host.Network().StopNotify(rm.connNotifee)
+		rm.closeSignal()
+		<-rm.closed
+		log.Infof("relay manager closed")
+	})
+	return nil
+}
+
+func (rm *RelayManager) keepHealthy() {
+	defer close(rm.closed)
+	for {
+		select {
+		case <-rm.closeCtx.Done():
+			log.Debugf("closing healthy checker")
+			return
+		case <-time.After(pollFrequency):
+			isProtected := rm.host.ConnManager().IsProtected(rm.relayAddr.ID, "")
+			connStatus := rm.host.Network().Connectedness(rm.relayAddr.ID)
+
+			if !isProtected || connStatus != network.Connected {
+				log.Warnf("detected unhealthy status of connection (protected: %t, connStatus: %s)", isProtected, connStatus)
+				if err := rm.connect(); err != nil {
+					log.Errorf("poller reconnect: %s", err)
+					continue
+				}
+			}
+			log.Debugf("relay connection is healthy")
+		}
+	}
+}
+
+func (rm *RelayManager) connect() error {
+	log.Infof("connecting with relay...")
+	err := rm.host.Connect(rm.closeCtx, rm.relayAddr)
+	if err != nil {
+		return fmt.Errorf("connecting to relay: %s", err)
+	}
+	rm.host.ConnManager().Protect(rm.relayAddr.ID, connProtectTag)
+	log.Infof("connected with relay")
+
+	return nil
+}
+
+type connNotifee struct {
+	rm *RelayManager
+}
+
+func (n *connNotifee) Disconnected(_ network.Network, ne network.Conn) {
+	if ne.RemotePeer() == n.rm.relayAddr.ID {
+		log.Warnf("disconnected from remote relay")
+		if err := n.rm.connect(); err != nil {
+			log.Errorf("notifee reconnect: %s", err)
+		}
+	}
+}
+func (n *connNotifee) Connected(_ network.Network, ne network.Conn)         {}
+func (n *connNotifee) OpenedStream(_ network.Network, s network.Stream)     {}
+func (n *connNotifee) ClosedStream(_ network.Network, s network.Stream)     {}
+func (n *connNotifee) ListenClose(_ network.Network, _ multiaddr.Multiaddr) {}
+func (n *connNotifee) Listen(net network.Network, ma multiaddr.Multiaddr)   {}
```

Comment:
The second healthiness mechanism is registering a `Notifee` (see L62), which the libp2p host will make calls to notify about connection events.

As soon as we detect a disconnect from the relay, we try firing a reconnection. If that succeeds that great since we reconnected as soon as that was detected. If that fails, we still rely on the other mechanism that in less than X seconds will detect this still is the case and reconnect (and keep trying every X secs).

We also print things in logs to signal the operator what's happening.